### ISSUE TITLE
How to mock the SecurityContext in a ?

### ISSUE THREAD

**Issue by gitfan13** [michaelyaakoby/testfun#15]
Dear Michael,

I'm trying to test the following WS:

```
	@GET
	@Path("abc")
	@Produces("application/json")
	public Abc abc(@Context SecurityContext sc) {
		final UserEntity user = userService.getByEmail(sc.getUserPrincipal().getName());
		return new Abc(user);
	}
```
In my unit test I'd like to get `sc.getUserPrincipal().getName()` to return a specific string, but a simple `@Mock` injects a different object than the one passed to the WS, so I'm getting a `NullPointerException` on `sc.getUserPrincipal()`. Here is my unit test:

```
@RunWith(EjbWithMockitoRunner.class)
public class AbcTest {
	@Mock
	private SecurityContext sc;

	@Test
	public void testAbc() {
		when(sc.getUserPrincipal()).thenReturn(new Principal() {
			@Override
			public String getName() {
				return "ab@c.de";
			}
		});

		jaxRsServer.jsonRequest("abc").expectStatus(Response.Status.OK).get());
	}
```

What's the proper way to do this? Is it even possible?

Thank you very much in advance!

**Comment by michaelyaakoby** [comment]
You need to mock the `SessionContext` (which contains the `SecurityContext`), see `MockSessionContextTest` in the readme file.

**Comment by gitfan13** [comment]
I saw that example, and I think that I understand it, but I'm still having trouble making the switch from `SessionContext` to `SecurityContext`.

Since you say that the `SessionContext` contains the `SecurityContext` I was expecting to find something like `sessionContext.getSecurityContext()`, which I could then mock, but I don't see it.

My unit test now contains `when(sessionContext.getCallerPrincipal()).thenReturn(new Principal() ...)` but my web service (which is working with **SecurityContext**) is still getting `null` when it calls `securityContext.getUserPrincipal()`.

**Comment by michaelyaakoby** [comment]
Got you. Indeed the current version doesn't inject `SecurityContext`.
Will try adding such support.

**Comment by gitfan13** [comment]
Perfect, thanks in advance :-)

Actually I think that the framework is already 90% there. When I debug the test execution of the `abc(@Context SecurityContext sc)` method of the web service under test, `sc` is a valid (non null) object. The only problem is that it's **not the same instance** as the one that I've created via `@Mock private SecurityContext sc;`

If I mock a service or a DAO in the unit test, you manage to inject the same instance in the class under test, so the technical support is already there. It just doesn't work for the SecurityContext.

If I can help in any way, please let me know: boji@swalk-app.com or Skype: bojidar-burgas

**Comment by michaelyaakoby** [comment]
In order to get SecuriyContext to contain a user-name, you must specify credentials in the HTTP request and configure the server with a security-domain.
I've pushed a change to the framework that adds such support. The test will look like:
```java
@RunWith(EjbWithMockitoRunner.class)
public class SecurityContextExampleTest {

    @Rule
    public JaxRsServer jaxRsServer = JaxRsServer.forResources(ExampleResource.class);

    @Test
    public void getUserFromSecurityContext() throws Exception {

        String response = jaxRsServer.
                jsonRequest("/example/user_from_security_context").
                basicAuth("kuki", "puki").
                expectStatus(Response.Status.OK).
                get();

        assertEquals("kuki", response);
    }
}
```

Note that I didn't yet publish a build with this change - let me know if this will work for you.

**Comment by gitfan13** [comment]
With the `basichAuth("username", "password")` method you are adding basic auth headers to the WS request, correct? But will the `jaxRsServer` properly process those headers (delegate the authentication and authorization to a class annotated with `@Provider` and `@PreMatching`) or simply ignore the password, fake a successful login and make `sc.getUserPrincipal()getName()` return "kuki"?

- A complete authentication and authorization would be awesome, but I assume that this is a lot of work for you.
- Simply faking a successful login and returning `kuki` would solve my current problem, but I think that there is a much more flexible solution, which is easy to implement: 
- Give us a way to provide a `java.util.Set` of Auth triplets to the server. A triplet contains a username, a password and a string with comma or space-separated user roles, and these would represent the valid users, their passwords and the user roles granted to each user. Maybe we can provide this Set like this:

```
@Rule
public JaxRsServer jaxRsServer = JaxRsServer.forResources(ExampleResource.class).**authData(...)**;
```

And now the `jaxRsServer` can easily validate the login data provided with the `basichAuth("username", "password")` method and return:

- 401 (Unauthorized) if the username/password combination is not contained in the authData set)
- 403 (Forbidden) if the granted roles do not contain the `@RolesAllowed("admin")` annotation of the web service method
- or execute the WS method and have the security context respond properly to `isUserInRole("abc")` and the Principal respond properly to `getName()`.

That would be awesome as it would allow us to unit test all the authentication and authorization logic contained in our WS or lower layers.

Again, I'm more than willing to help or get on Skype and chat.

**Comment by michaelyaakoby** [comment]
The short answer is yes... see my latest commit (change to infrastructure
and another example).
I've configured a security-domain which takes care of the authentication by
accepting any password you are sending. The rest is done by the jaxrs
server that is used by testfun.
I hope to release a new version with this capability sometime soon.
To contribute you can clone the project and then send a pull request.
Thanks,
M.

**Comment by gitfan13** [comment]
Your latest commit looks like it will solve my immediate problem, so thank you very much for that :-) After your release I'll do my first clone and pull request on GitHub with my proposed more flexible solution.

PS: please consider adding your proposed solution to issue #14 to your next release also.

**Comment by michaelyaakoby** [comment]
#14 is already in the previous release (minus ```withFormParams("p1=ABCD&p2=12345")```)

**Comment by michaelyaakoby** [comment]
All the above was released in 0.15.
